File: src/ctci6/utils/Queue.js
var Node = require('./Node');

function Queue() {
    this.N = 0;
    this.first = null;
    this.last = null;
}

Queue.prototype.add = function(item) {
    var node = new Node(item);
    if (!this.first) {
        this.first = node;
        this.last = node;
    } else {
        var last = this.last;
        last.next = node;
        this.last = node;
    }
    this.N++;
}

Queue.prototype.remove = function() {
    if (this.isEmpty()) return null;
    if (this.size() === 1) {
        var node = this.first;
        this.first = null;
        this.last = null;
        this.N--;
        return node.val;
    } else {
        var first = this.first;
        var second = this.first.next;
        this.first = second;
        this.N--;
        return first.val;
    }
}

Queue.prototype.first = function() {
    return this.first.val;
}

Queue.prototype.peek = function() {
    return this.last.val;
}

Queue.prototype.isEmpty = function() {
    return this.size() === 0;
}

Queue.prototype.size = function() {
    return this.N;
}

Queue.prototype.toString = function() {
    if (this.isEmpty()) return null;
    return this.first.toString();
}

module.exports = Queue;

function test() {
    var q = new Queue();
    q.add(1);
    q.add(2);
    q.remove();
    q.add(3);
    q.remove();
    q.remove();
    console.log(q.toString());
}

if (require.main === module) test();

File: ctci6/utils/Queue.java
package ctci6.utils;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by hao on 10/3/16.
 */
public class Queue<Value> implements Iterable<Value> {

    private class Node {
        private Value val;
        private Node next;
        public Node(Value val) {
            this.val = val;
        }
        public void setNext(Node next) {
            this.next = next;
        }
    }

    private Node first;
    private Node last;
    private int N;

    public Queue() {

    }

    public void enqueue(Value val) {
        Node node = new Node(val);
        if (first == null) {
            this.first = node;
            this.last = node;
        } else {
            Node oldLast = this.last;
            oldLast.setNext(node);
            this.last = node;
        }
        N++;
    }

    public Value dequeue() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }
        Node node = this.first;
        if (size() == 1) {
            first = null;
            last = null;
            N--;
            return node.val;
        } else {
            Node newFirst = node.next;
            first = newFirst;
            N--;
            return node.val;
        }
    }

    public int size() {
        return N;
    }

    public boolean isEmpty() {
        return N == 0;
    }

    public class QueueIterator implements Iterator {
        Node node = first;
        public Value next() {
            Node res = node;
            node = node.next;
            return res.val;
        }
        public boolean hasNext() {
            return node != null;
        }
    }

    public QueueIterator iterator() {
        return new QueueIterator();
    }

    public static void main(String[] args) {
        Queue<Integer> q = new Queue<>();
        q.enqueue(1);
        q.enqueue(2);
        for (Integer i : q) {
            System.out.println(i);
        }
    }

}
